### PR TITLE
Feat: add arrow to indicate live bus direction

### PR DESCRIPTION
**Pull request by marius-at-atb**
closes https://github.com/AtB-AS/kundevendt/issues/3688

busArrow2

Feature implemented in TravelDetailsMapScreenComponent with a new arrow from assets.

Mapbox had to be upgraded due to a bug in the original version. Thanks to @jorelosorio for great help with this!

Also note that there was a bug in React Native for Android using transform styles, which is why there is a special case for Android.

### CODE REVIEW

**Comment by marius-at-atb** [comment]
Please test all map functionality carefully because of the mapbox upgrade.
Also check Android carefully for rendering bugs of the arrow.

**Comment by gorandalum** [comment]
There is something strange here. Is the arrow icon code already merged to master?

**Comment by marius-at-atb** [comment]
@gorandalum Looks like it, but how? Was the upgrade-mapbox branch already merged to master at some point @jorelosorio, with these edits included?

**Comment by jorelosorio** [comment]
> @gorandalum Looks like it, but how? Was the upgrade-mapbox branch already merged to master at some point @jorelosorio, with these edits included?

This PR https://github.com/AtB-AS/mittatb-app/pull/3606 was initially pointing to master, and then I change to point your branch, but at some point I had to merge your branch into mine, so no sure why the changes ended up into master at the same time it was merged this into yours. It should not be on master.

**Comment by jorelosorio** [comment]
> > @gorandalum Looks like it, but how? Was the upgrade-mapbox branch already merged to master at some point @jorelosorio, with these edits included?
> 
> This PR #3606 was initially pointing to master, and then I change to point your branch, but at some point I had to merge your branch into mine, so no sure why the changes ended up into master at the same time it was merged this into yours. It should not be on master.

@gorandalum @marius-at-atb I do not know why it was merged into master but here is a revert of those changes, please approve it to come back to normal: https://github.com/AtB-AS/mittatb-app/pull/3634

File: src/travel-details-map-screen/TravelDetailsMapScreenComponent.tsx [hunk 1]
```diff
@@ -287,36 +292,39 @@ const LiveVehicle = ({
...
-              fill={
-                isError
-                  ? theme.interactive.interactive_destructive.default.background
-                  : isStale
-                  ? theme.interactive.interactive_1.default.background
-                  : fillColor
-              }
-              width={iconSize}
-              height={iconSize}
+          {pitch < 7 && ( // allow just a tiny bit of pitch before hiding the live bus direction arrow
```
Comment (KristianSelnas) [inline src/travel-details-map-screen/TravelDetailsMapScreenComponent.tsx:295]: Personally I would have found this code a lot more readable if this was extracted to a separate component.

Comment (marius-at-atb) [reply src/travel-details-map-screen/TravelDetailsMapScreenComponent.tsx:294]: Good point, will fix

Comment (marius-at-atb) [reply src/travel-details-map-screen/TravelDetailsMapScreenComponent.tsx:335]: @KristianSelnas See new commit for this one

File: src/travel-details-map-screen/TravelDetailsMapScreenComponent.tsx [hunk 2]
```diff
@@ -178,6 +180,47 @@ export const TravelDetailsMapScreenComponent = ({
   );
 };
 
+export type DirectionArrowParams = {
```
Comment (KristianSelnas) [inline src/travel-details-map-screen/TravelDetailsMapScreenComponent.tsx:183]: I believe the common convention is to call it Props, not Params, i.e. `DirectionArrowProps` (or possibly just `Props`, as it's only used locally in this one component).

File: src/travel-details-map-screen/TravelDetailsMapScreenComponent.tsx (hunk 1 above)
Comment (marius-at-atb) [reply src/travel-details-map-screen/TravelDetailsMapScreenComponent.tsx:335]: I have now separated DirectionArrow into its own file in an upcoming commit. Would you also want LiveVehicle and maybe even LiveVehicleIcon to be separated into their own files?

Comment (KristianSelnas) [reply src/travel-details-map-screen/TravelDetailsMapScreenComponent.tsx:333]: I can really just answer from a personal preference (which in most cases is "one component, one file). I think @gorandalum is better suited to answer what our official policy is. (I haven't checked if our coding guidelines mention this)

Comment (marius-at-atb) [reply src/travel-details-map-screen/TravelDetailsMapScreenComponent.tsx:294]: I just pushed the commit with DirectionArrow in its own file. Can make another one for LiveVehicle later if desirable.

**Review by KristianSelnas — APPROVED** [review]

File: src/travel-details-map-screen/TravelDetailsMapScreenComponent.tsx [hunk 3]
```diff
@@ -238,29 +250,64 @@ const LiveVehicle = ({
     };
   })();
 
-  if (!vehicle.location || zoomLevel < FOLLOW_MIN_ZOOM_LEVEL) return null;
+  if (
+    !vehicle.location ||
+    zoomLevel < FOLLOW_MIN_ZOOM_LEVEL ||
+    vehicle.bearing === undefined
```
Comment (jorelosorio) [inline src/travel-details-map-screen/TravelDetailsMapScreenComponent.tsx:256]: I was wondering if in the case that bearing (angle) is not provided, then we should show the bus only.

File: src/travel-details-map-screen/components/DirectionArrow.tsx [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  <View
+    style={{
+      shadowColor: '#000000',
+      shadowOffset: {width: 0, height: 2},
+      shadowOpacity: 0.2,
+      shadowRadius: 2,
+      position: 'absolute',
+      top: -Math.sin(bearingRadians) * directionArrowOffsetFromCenter,
+      left: Math.cos(bearingRadians) * directionArrowOffsetFromCenter,
+    }}
```
Comment (jorelosorio) [inline src/travel-details-map-screen/components/DirectionArrow.tsx:31]: I think we could move this to the style definition at the end of the file

**Review by gorandalum — CHANGES_REQUESTED** [review]
Good work! I have some comments, some on specific lines, and some written here:

- The direction arrow is also there when error (red) state. I think it should not be.
- When dragging the map and "untracking" the vehicle, the map centers on the bus again on updates. The logic revolving `shouldTrack` is ment to hinder this (not sure if it is introduced now or not). 
- Got this error in the console (not sure if it is introduced now or not):
> Mapbox error RCTMGLLayer.removeFromMap: removing layer failed for layer liveVehicleCircle: Layer liveVehicleCircle does not exist {"level": "error", "message": "RCTMGLLayer.removeFromMap: removing layer failed for layer liveVehicleCircle: Layer liveVehicleCircle does not exist"}

File: src/travel-details-map-screen/TravelDetailsMapScreenComponent.tsx [hunk 5]
```diff
@@ -238,29 +250,64 @@ const LiveVehicle = ({
...
+  )
+    return null;
+
+  const bearingRadians = ((90 - vehicle.bearing + heading) * Math.PI) / 180; // start at 90 degrees, go counter clockwise and convert from degrees to radians
+  const directionArrowOffsetFromCenter = 28;
+
+  const scaleForBugfix = Platform.OS === 'android' ? 2 : 1; // fix android transform rendering bugs by scaling up parent and child back down
```
[the remark this answers is not in the document]
Comment (marius-at-atb) [reply src/travel-details-map-screen/TravelDetailsMapScreenComponent.tsx:259]: Not a known bug as far as I can find? You can adjust the android scale factor to e.g. 1.5 to see the clipping bug.

**Comment by marius-at-atb** [comment]
> * When dragging the map and "untracking" the vehicle, the map centers on the bus again on updates. The logic revolving `shouldTrack` is ment to hinder this (not sure if it is introduced now or not).

@gorandalum  onTouchMove on MapboxGL.MapView doesn't fire.
Will use this on parent View instead:
```
onStartShouldSetResponder={() => true}
onResponderGrant={() => setShouldTrack(false)}
```

File: src/travel-details-map-screen/TravelDetailsMapScreenComponent.tsx [hunk 6]
```diff
@@ -116,14 +117,18 @@ export const TravelDetailsMapScreenComponent = ({
...
+        }}
       >
         <MapboxGL.Camera
           ref={mapCameraRef}
           bounds={bounds}
           {...MapCameraConfig}
-          zoomLevel={vehicleWithPosition ? FOLLOW_ZOOM_LEVEL : undefined}
-          centerCoordinate={vehicleWithPosition ? centerPosition : undefined}
+          zoomLevel={FOLLOW_ZOOM_LEVEL}
+          centerCoordinate={centerPosition}
```
[the remark this answers is not in the document]
Comment (marius-at-atb) [reply src/travel-details-map-screen/TravelDetailsMapScreenComponent.tsx]: I think not. Now reverted.

File: src/travel-details-map-screen/TravelDetailsMapScreenComponent.tsx [hunk 7]
```diff
@@ -116,14 +117,18 @@ export const TravelDetailsMapScreenComponent = ({
         onTouchMove={() => {
           setShouldTrack(false);
         }}
-        onRegionIsChanging={(event) => setZoomLevel(event.properties.zoomLevel)}
+        onCameraChanged={(state) => {
+          setHeading(state.properties.heading);
+          setZoomLevel(state.properties.zoom);
+          setPitch(state.properties.pitch);
```
[the remark this answers is not in the document]
Comment (marius-at-atb) [reply src/travel-details-map-screen/TravelDetailsMapScreenComponent.tsx]: Done in the new commit

**Review by jorelosorio — APPROVED** [review]
Good work, it was a smooth first issue experience

**Comment by marius-at-atb** [comment]
@jorelosorio Awesome help with the mapbox upgrade, thanks!
Very smooth issue indeed :P

**Review by gorandalum — COMMENTED** [review]
Good work! I will take a slight look at the `onTouchMove`, `onResponderGrant`, `onPressOut` logic before merging.

**Comment by marius-at-atb** [comment]
Elegant suggestion :+1:

**Pull request merged**